csrc/fusion_segmenter.cpp

```diff
@@ -316,16 +437,149 @@ std::unique_ptr<SegmentedFusion> SegmentedFusion::fromCompleteFusion(
   single_group->setHeuristic(heuristic);
   single_group->setID(0);
 
+  // Used to log the number of values and expressions in the fusion for
+  // serialization sanity check.
+  segmented_fusion_ptr->finalize();
   return segmented_fusion_ptr;
 }
 
 SegmentedFusion::SegmentedFusion(std::unique_ptr<Fusion> fusion)
     : segmented_fusion_name_{segmentedFusionName()},
       impl_(this),
-      complete_fusion_(std::move(fusion)) {
+      complete_fusion_(std::move(fusion)),
+      initial_vals_size_{0},
+      initial_exprs_size_{0} {
   annotateFP16IntermediateTensors();
 }
 
+flatbuffers::Offset<serde::SegmentedFusion> SegmentedFusion::serialize(
+    flatbuffers::FlatBufferBuilder& builder) const {
+  FUSER_PERF_SCOPE("SegmentedFusion::serialize");
+  const std::unordered_map<Val*, int64_t>& vals_map =
+      completeFusion()->deterministic_vals_map();
+  const std::unordered_map<Expr*, int64_t>& exprs_map =
+      completeFusion()->deterministic_exprs_map();
+  const std::unordered_map<SegmentedGroup*, int64_t>& groups_map =
+      impl_.deterministic_groups_map();
+  const std::unordered_map<SegmentedEdge*, int64_t>& edges_map =
+      impl_.deterministic_edges_map();
+
+  std::vector<flatbuffers::Offset<serde::SegmentedEdge>> edges_fb;
+  edges_fb.reserve(edges_.size());
+  for (auto se : edges_) {
+    edges_fb.push_back(serialize(builder, se, vals_map, groups_map));
+  }
+
+  std::vector<flatbuffers::Offset<serde::SegmentedGroup>> groups_fb;
+  groups_fb.reserve(groups_.size());
+  for (auto sg : groups_) {
+    groups_fb.push_back(
+        sg->serialize(builder, vals_map, exprs_map, groups_map, edges_map));
+  }
+
+  std::vector<int64_t> force_fp16_tv_fb;
+  force_fp16_tv_fb.reserve(force_fp16_tv_set_.size());
+  for (auto tv : force_fp16_tv_set_) {
+    force_fp16_tv_fb.push_back(vals_map.at(tv));
+  }
```

Comment:
I think this can be simplified using the more general `convertContainer` I suggested early on.